// @adonisjs/bodyparser/build/factories/main.js
import {
  defineConfig
} from "../chunk-IPABXONF.js";
import {
  BodyParserMiddleware
} from "../chunk-XVMZ4GI3.js";
import {
  MultipartFile
} from "../chunk-NRCULWNL.js";

// factories/file_factory.ts
var MultipartFileFactory = class {
  #parameters = {};
  /**
   * Merge factory params
   */
  merge(params) {
    this.#parameters = Object.assign(this.#parameters, params);
    return this;
  }
  /**
   * Create an instance of multipart file
   */
  create(validationOptions) {
    const file = new MultipartFile(
      {
        fieldName: this.#parameters.fieldName || "file",
        clientName: this.#parameters.clientName || this.#parameters.extname ? `file.${this.#parameters.extname}` : "file",
        headers: this.#parameters.headers || {}
      },
      validationOptions || {}
    );
    file.size = this.#parameters.size || 0;
    file.extname = this.#parameters.extname;
    file.type = this.#parameters.type;
    file.subtype = this.#parameters.subtype;
    file.state = "consumed";
    file.validate();
    return file;
  }
};

// factories/middleware_factory.ts
import lodash from "@poppinss/utils/lodash";
var BodyParserMiddlewareFactory = class {
  #config = defineConfig({});
  #getConfig() {
    return this.#config;
  }
  merge(config) {
    this.#config = lodash.merge(this.#config, config);
    return this;
  }
  create() {
    return new BodyParserMiddleware(this.#getConfig());
  }
};
export {
  BodyParserMiddlewareFactory,
  MultipartFileFactory
};
